src/log.cr

```diff
@@ -0,0 +1,128 @@
+#
+# When creating logs messages they belong to a _source_. If the top-level `Log` is used
+# as in the above examples its source is the empty string.
+#
+# The source can be used to identify the module or part of the application that is logging.
+# You can configure for each source a different level to filter the messages.
+#
+# A recommended patter is to declare a `Log` constant in the namespace of your shard or module as follows:
+#
+# ```
+# module DB
+#   Log = ::Log.for("db") # => Log for db source
+#
+#   def do_something
+#     Log.info { "this is logged in db source" }
+#   end
+# end
+#
+# DB::Log.info { "this is also logged in db source" }
+# Log.for("db").info { "this is also logged in db source" }
+# Log.info { "this is logged in top-level source" }
+# ```
+#
+# That way, any `Log.info` call within the `DB` module will use the `db` source. And not the top-level `::Log.info`.
+#
+# Sources can be nested. Continuing the last example, to declare a `Log` constant `db.pool` source you can do as follows:
+#
+# ```
+# class DB::Pool
+#   Log = DB::Log.for("pool") # => Log for db.pool source
+# end
+# ```
+#
+# A `Log` will emit the messages to the `Log::Backends` attached to it as long as
+# the configured severity filter `level` permits it.
+#
+# Logs can also be created from a type directly. The full name without generic arguments
+# will be used as a source. So for the type `Foo::Bar(Baz)` the source `foo.bar` will be used.
+#
+# ```
+# module DB
+#   Log = ::Log.for(self) # => Log for db source
+# end
+# ```
+#
+# ### Configure logging explicitly in the code
+#
+# Use `Log.builder` to indicate the which sources should go to which backends.
+#
+# You can indicate actual sources or patterns.
+#
+# * the empty string will show only the top-level source
+# * `*` will show all the sources
+# * `foo.bar.*` will show `foo.bar` and every nested source
+# * `foo.bar` will show `foo.bar`, but not its nested sources
+#
+#
+# The following configuration will setup for all sources to emit
+# warnings (or higher) to `STDOUT`, allow any of the `db.*` and
+# nested source to emit debug (or higher), and to also emit for all
+# sources errors (or higher) to an elasticsearch search.
+#
+# ```
+# backend = Log::StdioBackend.new
+# Log.builder.bind "*", :warning, backend
+# Log.builder.bind "db.*", :debug, backend
+# Log.builder.bind "*", :error, ElasticSearchBackend.new("http://localhost:9200")
+# ```
+#
+# ### Configure logging from environment variables
+#
+# Include the following line to allow configuration from environment variables.
+#
+# ```
+# Log.setup_from_env
+# ```
+#
+# The `CRYSTAL_LOG_LEVEL`, `CRYSTAL_LOG_SOURCES` environment variables are used to indicate
+# which severity level emit (defaults to `INFO`, use `NONE` to skip all messages) and to restrict
+# which sources you are interested in.
+#
+# The valid values for `CRYSTAL_LOG_SOURCES` are:
+#
+# * the empty string will show only the top-level source (default)
```

Comment:
Wouldn't it be better to refer to the above list and only clarify the default and that multiple values can be concatenated by comma?

Reply:
Can we have relative links in docs. Although the fact that the environment variable and the builder supports the same syntax is convenient, but they are different API.